lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,150 @@
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+ public:
+  SecretExtractToStaticExtractConversion(DataFlowSolver *solver,
+                                         MLIRContext *context)
+      : OpRewritePattern(context), solver(solver) {}
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO(#866): Add support for reads on multi-dimensional tensors
+    auto index = extractOp->getOperand(1);
+
+    auto *indexSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(index);
+
+    // Implies that the tensor.extract op is newly created by this pass
+    if (!indexSecretnessLattice) {
+      extractOp->emitWarning()
+          << "Secretness for tensor.extract index has not been set";
+      return failure();
+    }
+
+    bool isIndexSecret = indexSecretnessLattice->getValue().getSecretness();
+
+    // If index is not secret, no transformation is needed
+    if (!isIndexSecret) return failure();
+
+    ImplicitLocOpBuilder builder(extractOp->getLoc(), rewriter);
+
+    auto argType = extractOp.getResult().getType();
+
+    // Placeholder for extracted value
+    auto dummyValue = builder.create<arith::ConstantOp>(
+        argType, builder.getIntegerAttr(argType, 0));
+
+    auto sizeAttr = extractOp->getAttrOfType<IntegerAttr>("size");
+
+    // If size attribute is not provided, emit warning and return failure
+    if (!sizeAttr) {
+      extractOp->emitWarning()
+          << "Cannot convert secret tensor.extract to static tensor.extract "
+             "since a size attribute (`size`) has not "
+             "been provided on the tensor.extract op:";
+      return failure();
+    }
+
+    int size = sizeAttr.getInt();
+
+    SmallVector<Value> iterArgs = {dummyValue};
+    auto forOp = builder.create<affine::AffineForOp>(0, size, 1, iterArgs);
+
+    builder.setInsertionPointToStart(forOp.getBody());
+
+    // Check if the current index is equal to the secret index
+    auto cond = builder.create<arith::CmpIOp>(arith::CmpIPredicate::eq,
+                                              forOp.getInductionVar(), index);
+
+    // Extract value at current index
+    auto newExtractOp = builder.create<tensor::ExtractOp>(
+        extractOp->getOperand(0), forOp.getInductionVar());
+
+    auto ifOp = builder.create<scf::IfOp>(
+        cond,
+        [&](OpBuilder &b, Location loc) {
+          // Yield value extracted at index
+          b.create<scf::YieldOp>(loc, newExtractOp.getResult());
+        },
+        [&](OpBuilder &b, Location loc) {
+          // Yield dummyValue
+          b.create<scf::YieldOp>(loc, forOp.getRegionIterArgs().front());
+        });
+
+    // Create YieldOp for affine.for
+    SmallVector<Value> results(ifOp->getOpResults());
+    builder.create<affine::AffineYieldOp>(results);
+
+    // Replace the old tensor.insert op with forOp's result
+    rewriter.replaceOp(extractOp, forOp);
+
```

Comment:
Make sure to update the secretness of all new operations that you create.
For an example of how to do that, see https://github.com/google/heir/blob/f33691abaf3ad6dafb47a8f655ee630d8a45490a/lib/Transforms/ConvertIfToSelect/ConvertIfToSelect.cpp#L91-L96